### ISSUE TITLE
[Linux] Warning "libsecret_error: Failed to unlock the keyring"

### ISSUE THREAD

**Issue by PureTryOut** [juliansteenbakker/flutter_secure_storage#778]
I have an application that uses flutter_secure_storage to locally store credentials. Although I've ran this application successfully before it's been a while since I launched it and now it fails to unlock the keyring for some reason. I'm using KDE Plasma and it's kwallet is used as the keyring provider, with kwalletmanager I can still see the entry from last time I've used flutter_secure_storage so the keyring itself works just fine. I tried wiping the stored credentials and starting anew but it made no difference.

When reading any entry from the storage the console logs the following:

```
** (<appname>19565): WARNING **: 09:53:45.713: libsecret_error: Failed to unlock the keyring
```

It gives me no explanation on what exactly is failing so I have no way to figure out what's going wrong.

**Comment by PureTryOut** [comment]
When calling `.deleteAll()` it actually even errors out badly:

```
** (<appname>:27723): WARNING **: 15:41:52.952: libsecret_error: \xb2\xfe\u001d!
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 1)
#0      _Utf8Decoder.convertSingle (dart:convert-patch/convert_patch.dart:1760:7)
#1      Utf8Decoder.convert (dart:convert/utf.dart:349:37)
#2      StandardMessageCodec.readValueOfType (package:flutter/src/services/message_codecs.dart:502:29)
#3      StandardMessageCodec.readValue (package:flutter/src/services/message_codecs.dart:477:12)
#4      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:644:47)
#5      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:334:18)
<asynchronous suspension>
#6      FlutterSecureStorage.deleteAll (package:flutter_secure_storage/flutter_secure_storage.dart:285:5)
```

**Comment by PureTryOut** [comment]
This is still very much relevant and still breaks the whole functionality of this package on Linux!

**Comment by juliansteenbakker** [comment]
Sorry for the late response. Can you please check if this still an issue with version 10.0.0-beta.1? If so, I will try to setup a linux environment to check how i can solve this issue.

**Comment by SuricateTrading** [comment]
It still doesn't work in 10.0.0-beta.4.
I have a fresh Ubuntu installation, so it can't be me.

**Comment by gepbird** [comment]
Still relevant

**Comment by tomekit** [comment]
Can you guys provide what's your exact OS version and `flutter_secure_storage` version? 

It seems that this issue was initially reported in August 2024, yet `flutter_secure_storage` had its latest version, `9.2.4` released on Jan 9, 2025

**Comment by SuricateTrading** [comment]
> Can you guys provide what's your exact OS version and `flutter_secure_storage` version?
> 
> It seems that this issue was initially reported in August 2024, yet `flutter_secure_storage` had its latest version, `9.2.4` released on Jan 9, 2025

I use Ubuntu 24.04.2 LTS and flutter_secure_storage 9.2.4
So both versions are up to date.

**Comment by Unbreathable** [comment]
I use EndeavourOS (no version here because it's based on Arch) and flutter_secure_storage 9.2.4. I get the exact same problem.

I have a suggestion of how we could fix the issue though. There is a packaged called [``dbus_secrets``](https://pub.dev/packages/dbus_secrets) that works really well on Linux as it uses Canonical's ``dbus`` package to communicate directly with the ``org.freedesktop.secrets`` service. I've replaced my usage of ``flutter_secure_storage`` with ``dbus_secrets`` on Linux. However, the main package (as of writing this) has some issues on various desktops due to not following the standard completely. [I forked it](https://github.com/Liphium/dbus_secrets) and created a [pull request](https://github.com/akshaybabloo/dbus_secrets/pull/1) that makes it work in more places. After this, it at least works using ``gnome-keyring`` (default on Gnome) and ``kwallet`` (default on KDE), the only things I've tested it with so far.

If wanted, we could implement a new version of ``flutter_secure_storage_linux`` that uses ``dbus`` instead of the current approach with libsecret requiring native code.

**Comment by SuricateTrading** [comment]
A few weeks ago, it suddenly started working on Linux.
However, today I suddenly got the same error message again.
I fixed the problem by updating everything (operating system, packages, Dart, Flutter), and now it's working perfectly again.
This is really strange because when I first got the error message, I had a brand new Ubuntu installed, and everything was just installed and up to date.

**Comment by johnnyggalt** [comment]
I arrived here after attempting to add Linux support to my application (which already  runs on both Android and Windows).

I use Fedora Silverblue, but I think my findings will be of general interest.

I managed to get my app running without issue (including `flutter_secure_storage` usage) when building and running from within a Toolbox. I then moved onto solving the "IDE problem", by which I mean how can I connect my Flatpak-installed Code instance to the project-specific toolbox that contains the correct version of Flutter and other tooling?

I decided to [set up SSH](https://jurf.github.io/2020/03/24/vscode-toolbox/) in the toolbox and use Code's remote development tooling, along with the X-forwarding capabilities of SSH, to build and debug my app. I got that working but hit this issue (`Failed to unlock the keyring`). So it's interesting that it triggers when running the app within an SSH session, but not when running directly from the toolbox.

Anyway, I tried the suggestion from @Unbreathable (thanks!) and swapped out `flutter_secure_storage` for `dbus_secrets` and sure enough it worked. Of course, now I will need to introduce an abstraction in my app that either uses `dbus_secrets` on Linux or `flutter_secure_storage` for other platforms. It would be wonderful if this could be addressed in this library so I don't need this hoop-jumping.

**Comment by rangermix** [comment]
Still relevant
OS :Ubuntu 24.04.3 LTS on WSL2
flutter_secure_storage: both 9.2.4 and ^10.0.0-beta.4 would fail
hope that helps

```
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: PlatformException(Libsecret error, Failed to unlock the keyring, null, null)
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:653:7)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:367:18)
<asynchronous suspension>
#2      _SecureTokenStorage.load (package:personal_finance/services/token_storage_io.dart:18:16)
<asynchronous suspension>
#3      Session.loadFromStorage (package:personal_finance/services/session.dart:13:16)
<asynchronous suspension>
#4      _MyAppState.initState.<anonymous closure> (package:personal_finance/app.dart:23:7)
<asynchronous suspension>
```

**Comment by robbrit** [comment]
Hopefully this helps someone. For me the plugin was working fine on my Ubuntu desktop, but failing on my Debian CI machine (I run the integration test suite for my Flutter app as a Linux app in CI). The reason for this was that my code was running in a Docker container, it needed to unlock the keyring.

I fixed it by installing GNOME keyring in the Docker container:
* `apt install gnome-keyring` in the Dockerfile
* Launching the Docker container with `--cap-add ipc_lock`
* In the CI config that runs the tests:
```bash
dbus-run-session --bash -c "
  printf '\n' | gnome-keyring-daemon --unlock
  <command to run tests>
"
```
For those of you not running in Docker and are running the app on bare metal, perhaps you need to run `gnome-keyring-daemon --unlock` as part of your app startup in Linux.

**Comment by PureTryOut** [comment]
Your issue is different then, as I already had a working keyring installed, kwallet. I mentioned this in the original post. Please don't pollute the issue thread.

**Comment by tomekit** [comment]
> Your issue is different then, as I already had a working keyring installed, kwallet. I mentioned this in the original post. 

The thing is that you not only need to have keyring installed, but also unlocked. Thanks to @robbrit taking time to write his comment we now know how to unlock GNOME keyring and how `flutter_secure_storage` fails if you don't.

Apparently similar procedure is required for KDE Wallet, by calling: `/usr/lib/pam_kwallet_init` as mentioned here: https://wiki.archlinux.org/title/KDE_Wallet#Unlocking_KWallet_automatically_in_a_window_manager

> Please don't pollute the issue thread.

There is a possibility that you're experiencing a different problem, but given that other commenters are also experiencing this issue in other Linux distros and not just KDE I wouldn't even think about calling [helpful comment](https://github.com/juliansteenbakker/flutter_secure_storage/issues/778#issuecomment-3229525394) a pollution.

**Comment by SuricateTrading** [comment]
I've identified the cause and solution of this issue.

Ubuntu has a feature called “Automatic Login”, which lets you skip entering your user password each time you start your computer.
However, when this feature is enabled, the Keyring is not unlocked automatically at login.

Because of that, flutter_secure_storage (and any other library that relies on libsecret) cannot access the keyring — it remains locked, and the app itself has no way to unlock it since that requires the user’s Ubuntu password.

So, this isn’t actually a bug in flutter_secure_storage, but rather a side effect of Ubuntu’s login configuration.
That said, the error message could definitely be clearer and more helpful.


There are a few ways for users to fix this on their systems:

1. Disable automatic login and restart (recommended)
Go to Settings, search for "Users" and turn off “Automatic Login”

2. Manually unlock the keyring (inconvenient)

3. Use a keyring without a password (insecure)


In short, the issue is caused by the system configuration, not the app.
This can be problematic because many Ubuntu users have Automatic Login enabled by default, so they might all encounter the same confusing error message without understanding its cause.

**Comment by PureTryOut** [comment]
I don't use sutomatic login and my keyring _is_ automatically unlocked on login. I'm happy it works for you but it's not "the" cause or "the" solution.

**Comment by SuricateTrading** [comment]
What's the name of your keyrings? You should have a "login" keyring and I think it has to be set to default.

**Comment by Nieto2018** [comment]
> I've identified the cause and solution of this issue.
> 
> Ubuntu has a feature called “Automatic Login”, which lets you skip entering your user password each time you start your computer. However, when this feature is enabled, the Keyring is not unlocked automatically at login.
> 
> Because of that, flutter_secure_storage (and any other library that relies on libsecret) cannot access the keyring — it remains locked, and the app itself has no way to unlock it since that requires the user’s Ubuntu password.
> 
> So, this isn’t actually a bug in flutter_secure_storage, but rather a side effect of Ubuntu’s login configuration. That said, the error message could definitely be clearer and more helpful.
> 
> There are a few ways for users to fix this on their systems:
> 
> 1. Disable automatic login and restart (recommended)
>    Go to Settings, search for "Users" and turn off “Automatic Login”
> 2. Manually unlock the keyring (inconvenient)
> 3. Use a keyring without a password (insecure)
> 
> In short, the issue is caused by the system configuration, not the app. This can be problematic because many Ubuntu users have Automatic Login enabled by default, so they might all encounter the same confusing error message without understanding its cause.

Hi, I use Gnome Keyring with a "Login" entry without a password and the problem continues.

**Comment by kcdev001** [comment]
Not working in flatpak environment.

Config:
```
finish-args:
  - --socket=fallback-x11
  - --socket=wayland
  - --share=network
  - --socket=session-bus
  - --system-talk-name=org.freedesktop.NetworkManager
  - --talk-name=org.freedesktop.secrets
  - --talk-name=org.freedesktop.Secret.Service
  - --talk-name=org.gnome.keyring.*
  - --talk-name=org.freedesktop.Notifications
  - --filesystem=xdg-run/keyring
```



Error:
```
'/home/admin/snap/code/211/.local/share/flatpak/exports/share'

is not in the search path set by the XDG_DATA_DIRS environment variable, so
applications installed by Flatpak may not appear on your desktop until the
session is restarted.

Gtk-Message: 16:07:22.500: Failed to load module "canberra-gtk-module"
Gtk-Message: 16:07:22.505: Failed to load module "canberra-gtk-module"
libEGL warning: DRI3 error: Could not get DRI3 device
libEGL warning: Ensure your X server supports DRI3 to get accelerated rendering

(keychat:2): Atk-CRITICAL **: 16:07:23.048: atk_socket_embed: assertion 'plug_id != NULL' failed
IsarCore using libmdbx: v0.13.8-temp-upstream-fix

** (keychat:2): WARNING **: 16:07:23.495: libsecret_error: Failed to unlock the keyring
[ERROR:flutter/runtime/dart_vm_initializer.cc(40)] Unhandled Exception: PlatformException(Libsecret error, Failed to unlock the keyring, null, null)
#0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:653)
#1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:367)
<asynchronous suspension>
#2      SecureStorage.isBiometricsEnable (package:keychat/service/secure_storage.dart:176)
<asynchronous suspension>
```

**Comment by finnvyrn** [comment]
I met this issue when using Hyprland. It will occur when you use an uncommon display manager or not using a display manager. Following 2 to 4.1 of [GNOME/Keyring - ArchWiki](https://wiki.archlinux.org/title/GNOME/Keyring) solved my issue.

**Comment by jefcolbi** [comment]
I faced the same issue

flutter_secure_storage: ^10.0.0
Desktop manager: KDE
Ubuntu 20

`WARNING **: 14:15:55.776: libsecret_error: Failed to unlock the keyring`

My wallet is unlocked at login and i don't use an auto login feature.
More than 1 year for a bug.

**Comment by juliansteenbakker** [comment]
It's been a while but i've tried to fix some issues

> I have an application that uses flutter_secure_storage to locally store credentials. Although I've ran this application successfully before it's been a while since I launched it and now it fails to unlock the keyring for some reason. I'm using KDE Plasma and it's kwallet is used as the keyring provider, with kwalletmanager I can still see the entry from last time I've used flutter_secure_storage so the keyring itself works just fine. I tried wiping the stored credentials and starting anew but it made no difference.
> 
> When reading any entry from the storage the console logs the following:
> 
> ```
> ** (<appname>19565): WARNING **: 09:53:45.713: libsecret_error: Failed to unlock the keyring
> ```
> 
> It gives me no explanation on what exactly is failing so I have no way to figure out what's going wrong.

  The old implementation called secret_password_storev_sync with a NULL schema to "warm up" the keyring before every read. This was a workaround from Chromium (https://issues.chromium.org/issues/40490926) for a cold-keyring bug in gnome-keyring, but it seems to behave differently with KWallet and headless or SSH environments.

The warmup has been replaced with the proper libsecret service API: secret_service_get_sync + secret_collection_for_alias_sync + secret_service_unlock_sync. This follows the intended unlock flow, should work
  correctly with both gnome-keyring and KWallet backends, and no longer writes a dummy entry to your keyring on every read.

> When calling `.deleteAll()` it actually even errors out badly:
> 
> ```
> ** (<appname>:27723): WARNING **: 15:41:52.952: libsecret_error: \xb2\xfe\u001d!
> [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 1)
> #0      _Utf8Decoder.convertSingle (dart:convert-patch/convert_patch.dart:1760:7)
> #1      Utf8Decoder.convert (dart:convert/utf.dart:349:37)
> #2      StandardMessageCodec.readValueOfType (package:flutter/src/services/message_codecs.dart:502:29)
> #3      StandardMessageCodec.readValue (package:flutter/src/services/message_codecs.dart:477:12)
> #4      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:644:47)
> #5      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:334:18)
> <asynchronous suspension>
> #6      FlutterSecureStorage.deleteAll (package:flutter_secure_storage/flutter_secure_storage.dart:285:5)
> ```

deleteAll called storeToKeyring directly (bypassing the warmup), so when the keyring was locked or unavailable, the raw GError message from libsecret reached the Flutter method channel. This is fixed by sanitising the error message with g_utf8_make_valid before passing it to the channel.

Also,  deleteAll stored the JSON string "null" instead of "{}", because nlohmann::json() constructs a null value. This is not really clean, but shouldn't cause any issues afaik.

In order to make locked keyring issues debuggable, i've catched the generic "Libsecret error" and now thrown `PlatformException`,  so you can catch and handle that case explicitly like so:

 ```dart
 try {
    await storage.read(key: 'token');
  } on PlatformException catch (e) {
    if (e.code == 'KeyringLocked') {
      // Keyring is locked, prompt the user or handle gracefully
    }
  }
```

All of these changes are available via #1132. It would be great if someone can test it and give feedback on the changes. All tests are succesfull on ubuntu, but since this issue is caused on different distro's and setups, its hard to cover all possibilities.

**Comment by Nieto2018** [comment]
> It's been a while but i've tried to fix some issues
> 
> > I have an application that uses flutter_secure_storage to locally store credentials. Although I've ran this application successfully before it's been a while since I launched it and now it fails to unlock the keyring for some reason. I'm using KDE Plasma and it's kwallet is used as the keyring provider, with kwalletmanager I can still see the entry from last time I've used flutter_secure_storage so the keyring itself works just fine. I tried wiping the stored credentials and starting anew but it made no difference.
> > When reading any entry from the storage the console logs the following:
> > ```
> > ** (<appname>19565): WARNING **: 09:53:45.713: libsecret_error: Failed to unlock the keyring
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It gives me no explanation on what exactly is failing so I have no way to figure out what's going wrong.
> 
> The old implementation called secret_password_storev_sync with a NULL schema to "warm up" the keyring before every read. This was a workaround from Chromium (https://issues.chromium.org/issues/40490926) for a cold-keyring bug in gnome-keyring, but it seems to behave differently with KWallet and headless or SSH environments.
> 
> The warmup has been replaced with the proper libsecret service API: secret_service_get_sync + secret_collection_for_alias_sync + secret_service_unlock_sync. This follows the intended unlock flow, should work correctly with both gnome-keyring and KWallet backends, and no longer writes a dummy entry to your keyring on every read.
> 
> > When calling `.deleteAll()` it actually even errors out badly:
> > ```
> > ** (<appname>:27723): WARNING **: 15:41:52.952: libsecret_error: \xb2\xfe\u001d!
> > [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 1)
> > #0      _Utf8Decoder.convertSingle (dart:convert-patch/convert_patch.dart:1760:7)
> > #1      Utf8Decoder.convert (dart:convert/utf.dart:349:37)
> > #2      StandardMessageCodec.readValueOfType (package:flutter/src/services/message_codecs.dart:502:29)
> > #3      StandardMessageCodec.readValue (package:flutter/src/services/message_codecs.dart:477:12)
> > #4      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:644:47)
> > #5      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:334:18)
> > <asynchronous suspension>
> > #6      FlutterSecureStorage.deleteAll (package:flutter_secure_storage/flutter_secure_storage.dart:285:5)
> > ```
> 
> deleteAll called storeToKeyring directly (bypassing the warmup), so when the keyring was locked or unavailable, the raw GError message from libsecret reached the Flutter method channel. This is fixed by sanitising the error message with g_utf8_make_valid before passing it to the channel.
> 
> Also, deleteAll stored the JSON string "null" instead of "{}", because nlohmann::json() constructs a null value. This is not really clean, but shouldn't cause any issues afaik.
> 
> In order to make locked keyring issues debuggable, i've catched the generic "Libsecret error" and now thrown `PlatformException`, so you can catch and handle that case explicitly like so:
> 
> try {
>    await storage.read(key: 'token');
>  } on PlatformException catch (e) {
>    if (e.code == 'KeyringLocked') {
>      // Keyring is locked, prompt the user or handle gracefully
>    }
>  }
> All of these changes are available via [#1132](https://github.com/juliansteenbakker/flutter_secure_storage/pull/1132). It would be great if someone can test it and give feedback on the changes. All tests are succesfull on ubuntu, but since this issue is caused on different distro's and setups, its hard to cover all possibilities.

Hi @juliansteenbakker , I don't know if this will be of any help to you, but I've run the tests on my local machine with KDE Neon and the result is:

```
flutter drive --driver=test_driver/app_test_test.dart --target=integration_test/app_test.dart
Resolving dependencies... 
Downloading packages... 
  _fe_analyzer_shared 93.0.0 (100.0.0 available)
  analyzer 10.0.1 (13.0.0 available)
  matcher 0.12.19 (0.12.20 available)
  meta 1.17.0 (1.18.2 available)
  native_toolchain_c 0.17.6 (0.18.0 available)
  test 1.30.0 (1.31.1 available)
  test_api 0.7.10 (0.7.12 available)
  test_core 0.6.16 (0.6.18 available)
  vector_math 2.2.0 (2.3.0 available)
  very_good_analysis 7.0.0 (10.2.0 available)
Got dependencies!
10 packages have newer versions incompatible with dependency constraints.
Try `flutter pub outdated` for more information.
Building Linux application...                                           
✓ Built build/linux/x64/debug/bundle/flutter_secure_storage_example

** (flutter_secure_storage_example:192410): WARNING **: 11:28:54.311: Unable to retrieve framework response: Message is not valid JSON

** (flutter_secure_storage_example:192410): WARNING **: 11:28:54.470: Unable to retrieve framework response: Message is not valid JSON

** (flutter_secure_storage_example:192410): WARNING **: 11:28:54.490: Unable to retrieve framework response: Message is not valid JSON

** (flutter_secure_storage_example:192410): WARNING **: 11:28:54.590: Unable to retrieve framework response: Message is not valid JSON
VMServiceFlutterDriver: Connecting to Flutter application at http://127.0.0.1:37561/Lpdxb2uTiOI=/
VMServiceFlutterDriver: Isolate found with number: 5995508334643063
VMServiceFlutterDriver: Isolate 5995508334643063 is runnable.
VMServiceFlutterDriver: Isolate is paused at start.
VMServiceFlutterDriver: Attempting to resume isolate
00:00 +0: Secure Storage Tests Android: deleteAll() must not clear other sharedPreferencesName namespace (regression #1023)
00:00 +0 ~1: Secure Storage Tests Android: namespaces with different cipher algorithms must not interfere (full storageNamespace isolation)
00:00 +0 ~2: Secure Storage Tests Add a Random Row
VMServiceFlutterDriver: Connected to Flutter application.

** (flutter_secure_storage_example:192410): WARNING **: 11:28:56.187: Unable to retrieve framework response: Message is not valid JSON
00:03 +1 ~2: Secure Storage Tests Edit a Row Value
00:09 +2 ~2: Secure Storage Tests Delete a Row
00:12 +3 ~2: Secure Storage Tests Check Protected Data Availability
00:15 +4 ~2: Secure Storage Tests Contains Key for a Row
00:20 +5 ~2: Secure Storage Tests Read Value for a Row
00:28 +6 ~2: Secure Storage Tests Add Multiple Rows and Verify
00:31 +7 ~2: Secure Storage Tests Edit Multiple Rows
00:42 +8 ~2: Secure Storage Tests Delete All Rows
00:46 +9 ~2: Secure Storage Tests Enclave requested on iOS Simulator falls back gracefully
00:46 +9 ~3: Secure Storage Tests iOS device: baseline (useSecureEnclave=false) write/read/delete
00:46 +9 ~4: Secure Storage Tests iOS device: useSecureEnclave=true with non-prompting access control (applicationPassword) write/read/delete
00:46 +9 ~5: Secure Storage Tests iOS device: readAll with Secure Enclave items
00:46 +9 ~6: Secure Storage Tests iOS device: readAll with mixed Secure Enclave and standard items
00:46 +9 ~7: Secure Storage Tests Android: migrates single value from RSA_PKCS1/AES_CBC to OAEP/GCM
00:46 +9 ~8: Secure Storage Tests Android: migrates multiple values from RSA_PKCS1/AES_CBC to OAEP/GCM
00:46 +9 ~9: Secure Storage Tests Android: data remains readable without migration when algorithms unchanged
00:46 +9 ~10: Secure Storage Tests Android: migrateOnAlgorithmChange false skips migration
00:46 +9 ~11: Secure Storage Tests Android: migrateWithBackup migrates data from RSA_PKCS1/AES_CBC to OAEP/GCM
00:46 +9 ~12: Secure Storage Tests iOS device: item written without SE returns null when read with SE
00:46 +9 ~13: Secure Storage Tests iOS device: deleteAll with Secure Enclave items
00:46 +9 ~14: (tearDownAll)
Warning: integration_test plugin was not detected.

If you're running the tests with `flutter drive`, please make sure your tests
are in the `integration_test/` directory of your package and use
`flutter test $path_to_test` to run it instead.

If you're running the tests with Android instrumentation or XCTest, this means
that you are not capturing test results properly! See the following link for
how to set up the integration_test plugin:

https://docs.flutter.dev/testing/integration-tests

00:46 +10 ~14: All tests passed!
All tests passed.
```

**Comment by juliansteenbakker** [comment]
@Nieto2018 You should try running the specific linux tests like so: `flutter test integration_test/linux_test.dart -d linux`

**Comment by Nieto2018** [comment]
@juliansteenbakker the output with your command:

```
❯ flutter test integration_test/linux_test.dart -d linux
00:00 +0: loading /mnt/cosas/Escritorio/flutter_secure_storage/flutter_secure_storage/example/integration_test/linux_test.dart                                                                                             B00:18 +0: loading /mnt/cosas/Escritorio/flutter_secure_storage/flutter_secure_storage/example/integration_test/linux_test.dart                                                                                             
✓ Built build/linux/x64/debug/bundle/flutter_secure_storage_example
00:22 +15: All tests passed!  
```

**Comment by jefcolbi** [comment]
Good work

**Comment by johnnyggalt** [comment]
@juliansteenbakker Unfortunately, it still fails for my use case (see previous comment above).

If I run `flutter test integration_test/linux_test.dart -d linux` within my toolbox all tests pass. But if I SSH into the same toolbox and run the same command, I get a bunch of errors like this:

```
$ flutter test integration_test/linux_test.dart -d linux
00:00 +0: loading /var/home/met/repository/temp/flutter_secure_storage/flutter_secure_storage/example/integration_test/linux_test.dart
Building Linux application...
✓ Built build/linux/x64/debug/bundle/flutter_secure_storage_example
00:00 +0: Basic CRUD write and read round trip
00:00 +0 -1: Basic CRUD write and read round trip [E]
  FormatException: Invalid UTF-8 byte (at offset 0)
  dart:convert                                                                          Utf8Decoder.convert
  package:flutter/src/services/message_codecs.dart 502:29                               StandardMessageCodec.readValueOfType
  package:flutter/src/services/message_codecs.dart 477:12                               StandardMessageCodec.readValue
  package:flutter/src/services/message_codecs.dart 645:47                               StandardMethodCodec.decodeEnvelope
  package:flutter/src/services/platform_channel.dart 366:18                             MethodChannel._invokeMethod
  ===== asynchronous gap ===========================
  package:stream_channel                                                                _GuaranteeSink.add
  /home/me/tmp/flutter_tools.KIIXGG/flutter_test_listener.LNYERP/listener.dart 56:22  main.<fn>

  FormatException: Unexpected extension byte (at offset 0)
  dart:convert                                                                          Utf8Decoder.convert
  package:flutter/src/services/message_codecs.dart 502:29                               StandardMessageCodec.readValueOfType
  package:flutter/src/services/message_codecs.dart 477:12                               StandardMessageCodec.readValue
  package:flutter/src/services/message_codecs.dart 645:47                               StandardMethodCodec.decodeEnvelope
  package:flutter/src/services/platform_channel.dart 366:18                             MethodChannel._invokeMethod
  ===== asynchronous gap ===========================
  dart:async                                                                            _CustomZone.registerBinaryCallback
  integration_test/linux_test.dart 26:5                                                 main.<fn>
  ===== asynchronous gap ===========================
  package:stream_channel                                                                _GuaranteeSink.add
  /home/me/tmp/flutter_tools.KIIXGG/flutter_test_listener.LNYERP/listener.dart 56:22  main.<fn>
```

**Comment by juliansteenbakker** [comment]
> [@juliansteenbakker](https://github.com/juliansteenbakker) Unfortunately, it still fails for my use case (see previous comment above).
> 
> If I run `flutter test integration_test/linux_test.dart -d linux` within my toolbox all tests pass. But if I SSH into the same toolbox and run the same command, I get a bunch of errors like this:

@johnnyggalt I could actually replicate your error by running the example app without an accessible keyring daemon like so.

```shell
  cd flutter_secure_storage/example
  flutter build linux --debug
  dbus-run-session -- flutter test integration_test/linux_test.dart -d linux
```

The exception thrown was not catched. I've added the catch clause and now i get the following, which clearly states the keyring is locked.

```
  00:39 +0 -1: Basic CRUD write and read round trip [E]                                                                                                                                                             
    PlatformException(KeyringLocked, KeyringLocked, null, null)                                                                                                                                                     
    package:flutter/src/services/message_codecs.dart 653:7                      StandardMethodCodec.decodeEnvelope                                                                                                  
    package:flutter/src/services/platform_channel.dart 366:18                   MethodChannel._invokeMethod                                                                                                         
    ===== asynchronous gap ===========================                                                                                                                                                              
    package:stream_channel                                                      _GuaranteeSink.add                                                                                                                  
    /tmp/flutter_tools.BRBIFI/flutter_test_listener.SJOEKV/listener.dart 56:22  main.<fn>                                                                                                                           
                                                                                                                                                                                                                    
    PlatformException(KeyringLocked, KeyringLocked, null, null)                                                                                                                                                     
    package:flutter/src/services/message_codecs.dart 653:7                      StandardMethodCodec.decodeEnvelope                                                                                                  
    package:flutter/src/services/platform_channel.dart 366:18                   MethodChannel._invokeMethod                                                                                                         
    ===== asynchronous gap ===========================                                                                                                                                                              
    dart:async                                                                  _CustomZone.registerBinaryCallback                                                                                                  
    integration_test/linux_test.dart 26:5                                       main.<fn>                                                                                                                           
    ===== asynchronous gap ===========================                                                                                                                                                              
    package:stream_channel                                                      _GuaranteeSink.add                                                                                                                  
    /tmp/flutter_tools.BRBIFI/flutter_test_listener.SJOEKV/listener.dart 56:22  main.<fn>        
```

**Comment by johnnyggalt** [comment]
@juliansteenbakker Thanks, I can confirm that the error is now trapped. It doesn't really change the fact that it doesn't work for my scenario because the keyring is locked when connecting via SSH, even though using `dbus_secrets` _does_ work. I can't explain that and haven't dug deeper to figure out why. I realized that during development I can just use a simple "secrets" implementation that stores secrets in a flat file, bypassing the need for `dbus_secrets` and avoiding the keyring locked issue when using `flutter_secure_storage`.

**Comment by dp2022** [comment]
It's probably the same error i have on the Gnome. I've fixed it in the [PR 1163](https://github.com/juliansteenbakker/flutter_secure_storage/pull/1163). The main change is remove SECRET_SERVICE_LOAD_COLLECTIONS constant from the warmupKeyring function. This function also shouldn't "eat" errors and shouldn't try to open the keyring.

**Comment by EchoEllet** [comment]
> This was a workaround from Chromium (https://issues.chromium.org/issues/40490926) for a cold-keyring bug in gnome-keyring,

According to the issue report, it seems to be a GNOME libsecret API design inconsistency. It does not apply to the secret service API. Fortunately, PR #1182 rewrites `flutter_secure_storage_linux` to communicate with the D-Bus directly, without needing to work around libsecret.

I have never reproduced the libsecret issue myself. I only read the specification. [`package:freedesktop_secret`](https://pub.dev/packages/freedesktop_secret) handles unlocking collection/items and prompts to unlock the locked objects automatically.

To confirm this manually:

- Install `secret-tool` provided by GNOME libsecret (Ubuntu: `sudo apt install libsecret-tools`, Arch: `libsecret`)
- Run `secret-tool lock`
- Run the Flutter app and attempt any operation that calls any API that stores, reads, or deletes a secret (or run the integration tests)
- Expected: The prompt was handled as intended. The user can dismiss the dialog or complete the operation.
  - Dismissing the dialog throws a `SecretServicePromptDismissedException`. The library recommends [handling`Exception`](https://pub.dev/packages/freedesktop_secret#handle-thrown-exceptions).
  - If the prompt dialog was dismissed, the user already knows the context, so catching `Exception` instead of `SecretServicePromptDismissedException` and displaying a generic error message is acceptable.

When you say "cold keyring", I assume that's the same as locked collection?
According to the specification:

> A collection is similar in concept to the terms 'keyring' or 'wallet'.

This [integration test](https://github.com/juliansteenbakker/flutter_secure_storage/blob/develop/flutter_secure_storage/example/integration_test/linux_locked_keyring_test.dart#L26-L54) should probably be updated. 

The updated code attempts to prompt and only throws if the user explicitly dismissed the prompt. So `linux_locked_keyring_test.dart` can be updated accordingly.

> try {
   await storage.read(key: 'token');
 } on PlatformException catch (e) {
   if (e.code == 'KeyringLocked') {
     // Keyring is locked, prompt the user or handle gracefully
   }
 }

The updated implementation (PR #1182) no longer throws a `PlatformException` since it is meant for method channels. Clients that catch `Exception` will not be broken by this change. I would document this refactor in CHANGELOG.

**Comment by juliansteenbakker** [comment]
I think we can go ahead and close this issue since we closed an publish the following fixes:

  - #1223 / 3.0.3: skips warmupKeyring()'s D-Bus probe when sandboxed (Flatpak/Snap), which was the root cause of the cold/locked-keyring false positive for those environments.
  - v3.0.1: locked/unavailable keyring now surfaces as a catchable PlatformException(code: KeyringLocked) instead of an unstructured warning, and other libsecret failures come through as StorageError instead of malformed bytes.

For any other issues that are currently present, i would kindly ask to create new issues, since we otherwise lose track of it inside these long lived issues.
